File: flasher/device.py
import time
import shutil
import logging
from flasher.utils import run_cmd, fastboot, adb
import tempfile
from pathlib import Path
from dataclasses import dataclass, field

log = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

# Minimum acceptable transfer speed for cable health test (bytes/sec)
CABLE_SPEED_THRESHOLD_MB = 1.0

# Size of the temporary test payload written during cable speed test
CABLE_TEST_PAYLOAD_MB = 8

# How long to wait between device polling attempts (seconds)
POLL_INTERVAL = 3

# Total time to wait for a device to appear before giving up (seconds)
DEVICE_WAIT_TIMEOUT = 60

# Minimum recommended battery level before flashing
BATTERY_MIN_LEVEL = 30


# ---------------------------------------------------------------------------
# Data classes
# ---------------------------------------------------------------------------

@dataclass
class DeviceInfo:
    """Holds key variables retrieved from the device via fastboot getvar."""
    serial: str = ""
    product: str = ""
    variant: str = ""
    bootloader_version: str = ""
    baseband_version: str = ""
    secure: str = ""
    unlocked: str = ""
    battery_level: int = -1        # -1 = not reported
    slot_count: int = 1            # A/B devices report 2
    current_slot: str = ""
    raw: dict[str, str] = field(default_factory=dict)


@dataclass
class CableTestResult:
    passed: bool
    speed_mbs: float               # measured transfer speed in MB/s
    error: str = ""


@dataclass
class PreFlashCheck:
    device_found: bool
    communication_ok: bool
    cable_ok: bool
    battery_ok: bool
    unlocked: bool
    device_info: DeviceInfo = field(default_factory=DeviceInfo)
    cable_result: CableTestResult = field(default_factory=lambda: CableTestResult(False, 0.0))
    warnings: list[str] = field(default_factory=list)
    errors: list[str] = field(default_factory=list)

    @property
    def ready(self) -> bool:
        """True only when all hard requirements pass."""
        return (
            self.device_found
            and self.communication_ok
            and self.cable_ok
            and self.unlocked
        )


# ---------------------------------------------------------------------------
# Low-level fastboot / adb wrappers
# ---------------------------------------------------------------------------




# ---------------------------------------------------------------------------
# Device discovery
# ---------------------------------------------------------------------------

def list_fastboot_devices() -> list[str]:
    """Return serial numbers of all devices in fastboot or fastbootd mode."""
    rc, out, _ = fastboot("devices")
    if rc != 0 or not out:
        return []
    serials = []
    for line in out.splitlines():
        parts = line.split()
        # "fastboot devices" output:
        #   <serial>  fastboot   <- device in bootloader (old-style fastboot)
        #   <serial>  fastbootd  <- device in fastbootd (userspace fastboot)
        if len(parts) >= 2 and parts[1] in ("fastboot", "fastbootd"):
            serials.append(parts[0])
    return serials


def list_adb_devices() -> list[str]:
    """Return serial numbers of all devices reachable over adb."""
    rc, out, _ = adb("devices")
    if rc != 0:
        return []
    serials = []
    for line in out.splitlines()[1:]:   # skip the "List of devices" header
        parts = line.split()
        if len(parts) >= 2 and parts[1] == "device":
            serials.append(parts[0])
    return serials


def reboot_tofastbootd(serial: str | None = None) -> bool:
    """
    Reboot from ADB (system) directly into fastbootd via 'adb reboot fastboot'.

    Returns True if the adb command was accepted (device may still be
    rebooting — caller should poll with wait_for_device()).
    """
    cmd = ["-s", serial, "reboot", "fastboot"] if serial else ["reboot", "fastboot"]
    rc, _, err = adb(*cmd, timeout=15)
    if rc != 0:
        log.error(f"adb reboot fastboot failed: {err}")
        return False
    log.info("Reboot to fastbootd sent — waiting for device ...")
    return True


def wait_for_device(timeout: int = DEVICE_WAIT_TIMEOUT) -> str | None:
    """
    Poll fastboot devices until one appears or timeout is reached.

    Returns the serial number of the first device found, or None.
    """
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        serials = list_fastboot_devices()
        if serials:
            return serials[0]
        remaining = int(deadline - time.monotonic())
        log.info(f"No device found — retrying … ({remaining}s remaining)")
        time.sleep(POLL_INTERVAL)
    return None


# ---------------------------------------------------------------------------
# Device info
# ---------------------------------------------------------------------------

def _parse_getvar_output(output: str) -> dict[str, str]:
    """
    Parse 'fastboot getvar all' output into a key/value dict.

    fastboot prints variables to stderr in the form:
        (bootloader) key: value
    """
    result: dict[str, str] = {}
    for line in output.splitlines():
        line = line.strip().lstrip("(bootloader)").strip()
        if ":" in line:
            key, _, value = line.partition(":")
            result[key.strip()] = value.strip()
    return result


def get_device_info(serial: str | None = None) -> DeviceInfo | None:
    """
    Retrieve device variables via 'fastboot getvar all'.

    Returns None if communication fails.
    """
    cmd = ["getvar", "all"]
    if serial:
        cmd = ["-s", serial] + cmd

    # fastboot writes getvar output to stderr
    rc, stdout, stderr = fastboot(*cmd, timeout=15)
    raw_output = stderr or stdout

    if not raw_output:
        log.error("fastboot getvar all returned no output")
        return None

    raw = _parse_getvar_output(raw_output)

    def get(*keys: str) -> str:
        for k in keys:
            if k in raw:
                return raw[k]
        return ""

    # Battery level may be reported as "30" or "30%"
    battery_str = get("battery-level", "battery_level").rstrip("%")
    try:
        battery_level = int(battery_str)
    except ValueError:
        battery_level = -1

    try:
        slot_count = int(get("slot-count", "slot_count"))
    except ValueError:
        slot_count = 1

    return DeviceInfo(
        serial=serial or get("serialno"),
        product=get("product"),
        variant=get("variant"),
        bootloader_version=get("version-bootloader", "bootloader-version"),
        baseband_version=get("version-baseband", "baseband-version"),
        secure=get("secure"),
        unlocked=get("unlocked"),
        battery_level=battery_level,
        slot_count=slot_count,
        current_slot=get("current-slot"),
        raw=raw,
    )


# ---------------------------------------------------------------------------
# Cable speed test
# ---------------------------------------------------------------------------

def test_cable_speed(serial: str | None = None) -> CableTestResult:
    """
    Estimate USB transfer speed using 'fastboot stage' (RAM only, no NAND write).

    fastboot stage uploads a file into device RAM without touching any partition.
    Falls back to flashing 'tmp' partition if stage is not supported.
    """
    payload_bytes = CABLE_TEST_PAYLOAD_MB * 1024 * 1024

    with tempfile.NamedTemporaryFile(suffix=".img", delete=False) as f:
        f.write(b"\x00" * payload_bytes)
        tmp_path = f.name

    serial_args = ["-s", serial] if serial else []

    try:
        start = time.monotonic()
        # 'fastboot stage' uploads to RAM only — safe on all devices
        rc, _, err = fastboot(*serial_args, "stage", tmp_path, timeout=60)
        elapsed = time.monotonic() - start

        if rc != 0:
            # stage not supported — skip speed test rather than risk data loss
            return CableTestResult(
                passed=True,
                speed_mbs=0.0,
                error="fastboot stage not supported on this device (speed test skipped)",
            )
    finally:
        Path(tmp_path).unlink(missing_ok=True)

    if elapsed <= 0:
        return CableTestResult(passed=False, speed_mbs=0.0, error="Elapsed time was zero")

    speed_mbs = payload_bytes / elapsed / (1024 * 1024)
    passed = speed_mbs >= CABLE_SPEED_THRESHOLD_MB

    if not passed:
        log.warning(
            f"Cable speed {speed_mbs:.2f} MB/s is below threshold "
            f"({CABLE_SPEED_THRESHOLD_MB} MB/s) — try a different cable or port"
        )
    else:
        log.info(f"Cable speed OK: {speed_mbs:.2f} MB/s")

    return CableTestResult(passed=passed, speed_mbs=speed_mbs)


# ---------------------------------------------------------------------------
# Pre-flash check orchestrator
# ---------------------------------------------------------------------------

def run_pre_flash_checks(serial: str | None = None) -> PreFlashCheck:
    """
    Run all pre-flash diagnostics and return a PreFlashCheck summary.

    Steps:
      1. Detect device via fastboot devices
      2. Attempt adb reboot if not found
      3. Retrieve device info (fastboot getvar all)
      4. Test cable transfer speed
      5. Validate battery level and bootloader unlock state
    """
    check = PreFlashCheck(
        device_found=False,
        communication_ok=False,
        cable_ok=False,
        battery_ok=False,
        unlocked=False,
    )

    # ------------------------------------------------------------------
    # Step 1 — device discovery
    # ------------------------------------------------------------------
    log.info("==> [1/4] Detecting device …")
    serials = list_fastboot_devices()

    if not serials:
        # Not in fastboot/fastbootd — check adb (device in system)
        adb_serials = list_adb_devices()

        if adb_serials:
            adb_serial = serial or adb_serials[0]
            log.info(f"Device found in system mode via adb ({adb_serial})")
            log.info("Pre-flash checks will run in adb mode — reboot handled separately")
            # Mark as found but skip fastboot-specific checks
            check.device_found = True
            check.communication_ok = True  # adb is working
            check.cable_ok = True          # cable is fine if adb works
            check.battery_ok = True        # skip battery check
            check.unlocked = True          # cannot check unlock from adb, assume ok
            return check
        else:
            check.errors.append(
                "No device found via fastboot or adb. "
                "Boot the device into fastboot manually: hold Vol Down + Power"
            )
            return check

    serial = serial or serials[0]
    check.device_found = True
    log.info(f"Device found: {serial}")

    # ------------------------------------------------------------------
    # Step 2 — communication test (fastboot getvar all)
    # ------------------------------------------------------------------
    log.info("==> [2/4] Testing communication (fastboot getvar all) …")
    info = get_device_info(serial)

    if info is None:
        check.errors.append("fastboot getvar all failed — device may be unresponsive")
        return check

    check.communication_ok = True
    check.device_info = info

    log.info(f"  Product       : {info.product}")
    log.info(f"  Variant       : {info.variant}")
    log.info(f"  Bootloader    : {info.bootloader_version}")
    log.info(f"  Baseband      : {info.baseband_version}")
    log.info(f"  Secure boot   : {info.secure}")
    log.info(f"  Unlocked      : {info.unlocked}")
    log.info(f"  Battery       : {info.battery_level}%")
    if info.slot_count == 2:
        log.info(f"  A/B device    : current slot = {info.current_slot}")

    # ------------------------------------------------------------------
    # Step 3 — cable speed test
    # ------------------------------------------------------------------
    log.info("==> [3/4] Testing cable transfer speed …")
    cable = test_cable_speed(serial)
    check.cable_result = cable
    check.cable_ok = cable.passed

    if not cable.passed:
        check.warnings.append(
            f"Slow cable detected ({cable.speed_mbs:.2f} MB/s). "
            "Flashing may take longer or fail. Use a USB 3.0 cable and port."
        )

    # ------------------------------------------------------------------
    # Step 4 — safety checks
    # ------------------------------------------------------------------
    log.info("==> [4/4] Running safety checks …")

    # Bootloader unlock
    if info.unlocked.lower() in ("yes", "true", "1"):
        check.unlocked = True
    else:
        check.errors.append(
            "Bootloader is locked. Unlock it first: fastboot flashing unlock"
        )

    # Battery level
    if info.battery_level == -1:
        check.warnings.append("Battery level not reported by device")
        check.battery_ok = True   # can't block on unknown value
    elif info.battery_level < BATTERY_MIN_LEVEL:
        check.battery_ok = False
        check.errors.append(
            f"Battery too low ({info.battery_level}%). "
            f"Charge to at least {BATTERY_MIN_LEVEL}% before flashing."
        )
    else:
        check.battery_ok = True
        log.info(f"  Battery level OK: {info.battery_level}%")

    return check


# ---------------------------------------------------------------------------
# CLI — standalone diagnostics runner
# ---------------------------------------------------------------------------

def _print_check_report(check: PreFlashCheck) -> None:
    """Print a human-readable summary of PreFlashCheck results."""
    def status(ok: bool) -> str:
        return "✓" if ok else "✗"

    print("\n── Pre-flash check report ─────────────────────────────")
    print(f"  {status(check.device_found):<3} Device detected")
    print(f"  {status(check.communication_ok):<3} Fastboot communication")
    print(f"  {status(check.cable_ok):<3} Cable speed "
          f"({check.cable_result.speed_mbs:.2f} MB/s)")
    print(f"  {status(check.battery_ok):<3} Battery level "
          f"({check.device_info.battery_level}%)")
    print(f"  {status(check.unlocked):<3} Bootloader unlocked")

    if check.warnings:
        print("\n  Warnings:")
        for w in check.warnings:
            print(f"    ⚠  {w}")

    if check.errors:
        print("\n  Errors:")
        for e in check.errors:
            print(f"    ✗  {e}")

    print()
    if check.ready:
        print("  ✓ Device is ready for flashing.")
    else:
        print("  ✗ Device is NOT ready. Fix the errors above before flashing.")
    print("────────────────────────────────────────────────────────\n")


def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(
        prog="device",
        description="Run pre-flash diagnostics on a connected Android device",
    )
    parser.add_argument(
        "-s", "--serial",
        default=None,
        metavar="SERIAL",
        help="Target a specific device by serial number",
    )
    args = parser.parse_args()

    # Verify fastboot is available
    if not shutil.which("fastboot"):
        print("✗ fastboot not found in $PATH — install android-tools")
        raise SystemExit(1)

    check = run_pre_flash_checks(serial=args.serial)
    _print_check_report(check)

    raise SystemExit(0 if check.ready else 1)


if __name__ == "__main__":
    main()

File: lib/src/device.rs
//! Device discovery and pre-flash diagnostics.
//!
//! Provides device detection (fastboot/adb), info retrieval via `fastboot getvar`,
//! cable speed testing via `fastboot stage`, and a comprehensive pre-flash check.

use std::collections::HashMap;
use std::fs;
use std::thread;
use std::time::{Duration, Instant};

use tracing::{error, info, warn};

use crate::utils::run_cmd;

const CABLE_SPEED_THRESHOLD_MB: f64 = 1.0;
const CABLE_TEST_PAYLOAD_MB: usize = 8;
/// Device polling interval in milliseconds. `fastboot devices` is a cheap
/// local USB enumeration, so polling twice a second keeps detection snappy.
const POLL_INTERVAL_MS: u64 = 500;
const BATTERY_MIN_LEVEL: i32 = 30;

// ---------------------------------------------------------------------------
// Data structures
// ---------------------------------------------------------------------------

/// Key device variables retrieved via fastboot getvar.
#[derive(Debug, Clone, Default)]
pub struct DeviceInfo {
    pub serial: String,
    pub product: String,
    pub variant: String,
    pub bootloader_version: String,
    pub baseband_version: String,
    pub secure: String,
    pub unlocked: String,
    pub battery_level: i32, // -1 = not reported
    pub slot_count: i32,    // A/B devices report 2
    pub current_slot: String,
    pub raw: HashMap<String, String>,
}

impl DeviceInfo {
    pub fn new() -> Self {
        Self {
            battery_level: -1,
            slot_count: 1,
            ..Default::default()
        }
    }
}

#[derive(Debug, Clone, Default)]
pub struct CableTestResult {
    pub passed: bool,
    pub speed_mbs: f64,
    pub error: String,
}

/// Comprehensive pre-flash check result.
#[derive(Debug, Clone)]
pub struct PreFlashCheck {
    pub device_found: bool,
    pub communication_ok: bool,
    pub cable_ok: bool,
    pub battery_ok: bool,
    pub unlocked: bool,
    pub device_info: DeviceInfo,
    pub cable_result: CableTestResult,
    pub warnings: Vec<String>,
    pub errors: Vec<String>,
}

impl Default for PreFlashCheck {
    fn default() -> Self {
        Self::new()
    }
}

impl PreFlashCheck {
    pub fn new() -> Self {
        Self {
            device_found: false,
            communication_ok: false,
            cable_ok: false,
            battery_ok: false,
            unlocked: false,
            device_info: DeviceInfo::new(),
            cable_result: CableTestResult::default(),
            warnings: Vec::new(),
            errors: Vec::new(),
        }
    }
    /// True only when all hard requirements pass.
    pub fn ready(&self) -> bool {
        self.device_found
            && self.communication_ok
            && self.cable_ok
            && self.battery_ok
            && self.unlocked
    }
}

// ---------------------------------------------------------------------------
// Device discovery
// ---------------------------------------------------------------------------

/// Return serial numbers of devices in fastboot/fastbootd mode.
pub fn list_fastboot_devices() -> Vec<String> {
    let r = run_cmd(&["fastboot", "devices"], 5);
    if r.code != 0 || r.stdout.is_empty() {
        return Vec::new();
    }
    r.stdout
        .lines()
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 && (parts[1] == "fastboot" || parts[1] == "fastbootd") {
                Some(parts[0].to_string())
            } else {
                None
            }
        })
        .collect()
}

/// Return serial numbers of devices in fastbootd mode only.
pub fn list_fastbootd_devices() -> Vec<String> {
    let r = run_cmd(&["fastboot", "devices"], 5);
    if r.code != 0 || r.stdout.is_empty() {
        return Vec::new();
    }
    r.stdout
        .lines()
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 && parts[1] == "fastbootd" {
                Some(parts[0].to_string())
            } else {
                None
            }
        })
        .collect()
}

/// Return serial numbers of devices reachable via adb.
pub fn list_adb_devices() -> Vec<String> {
    let r = run_cmd(&["adb", "devices"], 5);
    if r.code != 0 {
        return Vec::new();
    }
    r.stdout
        .lines()
        .skip(1)
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2 && parts[1] == "device" {
                Some(parts[0].to_string())
            } else {
                None
            }
        })
        .collect()
}

/// Reboot device from ADB into fastbootd.
pub fn reboot_to_fastbootd(serial: Option<&str>) -> bool {
    let mut cmd: Vec<&str> = vec!["adb"];
    let s;
    if let Some(ser) = serial {
        s = ser.to_string();
        cmd.push("-s");
        cmd.push(&s);
    }
    cmd.extend(&["reboot", "fastboot"]);
    let r = run_cmd(&cmd, 15);
    if r.code != 0 {
        error!("adb reboot fastboot failed: {}", r.stderr);
        return false;
    }
    info!("Reboot to fastbootd sent — waiting for device ...");
    true
}

/// Poll fastboot devices until one appears or timeout expires.
pub fn wait_for_device(timeout: u64) -> Option<String> {
    let deadline = Instant::now() + Duration::from_secs(timeout);
    let mut last_log = Instant::now();
    while Instant::now() < deadline {
        let serials = list_fastboot_devices();
        if !serials.is_empty() {
            return serials.into_iter().next();
        }
        if last_log.elapsed().as_secs() >= 5 {
            let remaining = (deadline - Instant::now()).as_secs();
            info!("No device found — retrying … ({}s remaining)", remaining);
            last_log = Instant::now();
        }
        thread::sleep(Duration::from_millis(POLL_INTERVAL_MS));
    }
    None
}

// ---------------------------------------------------------------------------
// Device info
// ---------------------------------------------------------------------------

/// Parse `fastboot getvar all` output into key/value pairs.
fn parse_getvar_output(output: &str) -> HashMap<String, String> {
    let mut result = HashMap::new();
    for line in output.lines() {
        let line = line.trim().trim_start_matches("(bootloader)").trim();
        if let Some((key, value)) = line.split_once(':') {
            result.insert(key.trim().to_string(), value.trim().to_string());
        }
    }
    result
}

/// Retrieve device info via `fastboot getvar all`. Returns None on failure.
pub fn get_device_info(serial: Option<&str>) -> Option<DeviceInfo> {
    let mut cmd: Vec<&str> = vec!["fastboot"];
    let s;
    if let Some(ser) = serial {
        s = ser.to_string();
        cmd.push("-s");
        cmd.push(&s);
    }
    cmd.extend(&["getvar", "all"]);

    let r = run_cmd(&cmd, 15);
    let raw_output = if r.stderr.is_empty() {
        &r.stdout
    } else {
        &r.stderr
    };
    if raw_output.is_empty() {
        error!("fastboot getvar all returned no output");
        return None;
    }

    let raw = parse_getvar_output(raw_output);
    let get = |keys: &[&str]| -> String {
        for k in keys {
            if let Some(v) = raw.get(*k) {
                return v.clone();
            }
        }
        String::new()
    };

    let battery = get(&["battery-level", "battery_level"])
        .trim_end_matches('%')
        .parse::<i32>()
        .unwrap_or(-1);
    let slots = get(&["slot-count", "slot_count"])
        .parse::<i32>()
        .unwrap_or(1);

    Some(DeviceInfo {
        serial: serial
            .map(|s| s.to_string())
            .unwrap_or_else(|| get(&["serialno"])),
        product: get(&["product"]),
        variant: get(&["variant"]),
        bootloader_version: get(&["version-bootloader", "bootloader-version"]),
        baseband_version: get(&["version-baseband", "baseband-version"]),
        secure: get(&["secure"]),
        unlocked: get(&["unlocked"]),
        battery_level: battery,
        slot_count: slots,
        current_slot: get(&["current-slot"]),
        raw,
    })
}

// ---------------------------------------------------------------------------
// Cable speed test
// ---------------------------------------------------------------------------

/// Estimate USB transfer speed via `fastboot stage` (RAM only, no NAND write).
pub fn test_cable_speed(serial: Option<&str>) -> CableTestResult {
    let bytes = CABLE_TEST_PAYLOAD_MB * 1024 * 1024;
    // Unpredictable per-run temp file: a fixed /tmp name is open to symlink
    // attacks and collides between concurrent runs. Deleted on drop.
    let tmp = match tempfile::Builder::new()
        .prefix("lfff-cable-")
        .suffix(".img")
        .tempfile()
    {
        Ok(f) => f,
        Err(e) => {
            return CableTestResult {
                passed: false,
                speed_mbs: 0.0,
                error: format!("Cannot create test payload: {}", e),
            };
        }
    };
    if let Err(e) = fs::write(tmp.path(), vec![0u8; bytes]) {
        return CableTestResult {
            passed: false,
            speed_mbs: 0.0,
            error: format!("Cannot create test payload: {}", e),
        };
    }

    let ts = tmp.path().to_string_lossy().to_string();
    let mut cmd: Vec<&str> = vec!["fastboot"];
    let s;
    if let Some(ser) = serial {
        s = ser.to_string();
        cmd.push("-s");
        cmd.push(&s);
    }
    cmd.extend(&["stage", &ts]);

    let start = Instant::now();
    let r = run_cmd(&cmd, 60);
    let elapsed = start.elapsed().as_secs_f64();
    drop(tmp);

    if r.code != 0 {
        let stderr_lower = r.stderr.to_lowercase();
        let is_unsupported = stderr_lower.contains("unknown command")
            || stderr_lower.contains("not supported")
            || stderr_lower.contains("not implemented")
            || stderr_lower.contains("no such command");
        if is_unsupported {
            return CableTestResult {
                passed: true,
                speed_mbs: 0.0,
                error: "fastboot stage not supported on this device (speed test skipped)".into(),
            };
        }
        return CableTestResult {
            passed: false,
            speed_mbs: 0.0,
            error: format!("Cable speed test failed: {}", r.stderr),
        };
    }
    if elapsed <= 0.0 {
        return CableTestResult {
            passed: false,
            speed_mbs: 0.0,
            error: "Elapsed time was zero".into(),
        };
    }

    let speed = bytes as f64 / elapsed / (1024.0 * 1024.0);
    let passed = speed >= CABLE_SPEED_THRESHOLD_MB;
    if !passed {
        warn!("Cable speed {:.2} MB/s is below threshold", speed);
    } else {
        info!("Cable speed OK: {:.2} MB/s", speed);
    }
    CableTestResult {
        passed,
        speed_mbs: speed,
        error: String::new(),
    }
}

// ---------------------------------------------------------------------------
// Pre-flash check orchestrator
// ---------------------------------------------------------------------------

/// Run all pre-flash diagnostics and return a summary.
pub fn run_pre_flash_checks(serial: Option<&str>) -> PreFlashCheck {
    let mut c = PreFlashCheck::new();

    // Step 1: device discovery
    info!("==> [1/4] Detecting device …");
    let serials = list_fastboot_devices();
    if serials.is_empty() {
        let adb_s = list_adb_devices();
        if !adb_s.is_empty() {
            c.device_found = true;
            c.errors.push(
                "Device found via ADB, but flashing requires fastboot/fastbootd mode. \
                 Please reboot to bootloader (Vol Down + Power)."
                    .into(),
            );
            return c;
        }
        c.errors.push(
            "No device found via fastboot or adb. Boot into fastboot: hold Vol Down + Power".into(),
        );
        return c;
    }

    let ser = serial
        .map(|s| s.to_string())
        .unwrap_or_else(|| serials[0].clone());
    c.device_found = true;
    info!("Device found: {}", ser);

    // Step 2: communication
    info!("==> [2/4] Testing communication …");
    let di = match get_device_info(Some(&ser)) {
        Some(i) => i,
        None => {
            c.errors.push("fastboot getvar all failed".into());
            return c;
        }
    };
    c.communication_ok = true;
    c.device_info = di.clone();
    info!(
        "  Product: {} | Unlocked: {} | Battery: {}%",
        di.product, di.unlocked, di.battery_level
    );

    // Step 3: cable speed
    info!("==> [3/4] Testing cable speed …");
    let cable = test_cable_speed(Some(&ser));
    c.cable_ok = cable.passed;
    if !cable.passed {
        // ready() treats a failed cable test as a hard error, so report it in
        // errors — otherwise the user sees "checks failed" with no reason listed.
        if cable.error.is_empty() {
            c.errors.push(format!(
                "Cable too slow ({:.2} MB/s, need ≥{:.1} MB/s). Use a different cable or a USB 3.0 port.",
                cable.speed_mbs, CABLE_SPEED_THRESHOLD_MB
            ));
        } else {
            c.errors.push(cable.error.clone());
        }
    }
    c.cable_result = cable;

    // Step 4: safety
    info!("==> [4/4] Safety checks …");
    let ul = di.unlocked.to_lowercase();
    if ul == "yes" || ul == "true" || ul == "1" {
        c.unlocked = true;
    } else {
        c.errors
            .push("Bootloader is locked. Unlock: fastboot flashing unlock".into());
    }

    if di.battery_level == -1 {
        c.warnings.push("Battery level not reported".into());
        c.battery_ok = true;
    } else if di.battery_level < BATTERY_MIN_LEVEL {
        c.battery_ok = false;
        c.errors.push(format!(
            "Battery too low ({}%). Need ≥{}%.",
            di.battery_level, BATTERY_MIN_LEVEL
        ));
    } else {
        c.battery_ok = true;
    }

    c
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_getvar() {
        let out = "(bootloader) product: lemonade\n(bootloader) unlocked: yes\n(bootloader) battery-level: 85%";
        let v = parse_getvar_output(out);
        assert_eq!(v.get("product"), Some(&"lemonade".to_string()));
        assert_eq!(v.get("unlocked"), Some(&"yes".to_string()));
    }

    #[test]
    fn test_preflash_ready() {
        let mut c = PreFlashCheck::new();
        c.device_found = true;
        c.communication_ok = true;
        c.cable_ok = true;
        c.battery_ok = true;
        c.unlocked = true;
        assert!(c.ready());
        c.battery_ok = false;
        assert!(!c.ready());
        c.battery_ok = true;
        c.unlocked = false;
        assert!(!c.ready());
    }
}
